File: python/algovoi_substrate/audit_chain.py
"""
Audit chain primitive: monotonic per-row hash chain.

Each row carries:
- content_hash: SHA-256(JCS(payload)) -- the per-row substrate hash.
- prev_hash: the previous row's content_hash, or null for the head row.
- chain_position: monotonically increasing integer starting at 0.

This is the structure AlgoVoi's production /compliance/attestation chain
exposes, with Object Lock COMPLIANCE retention to 2033-05-04. A year-five
supervisor reading retained bytes can recompute content_hash from the
payload + canon_version pin and verify prev_hash linkage walks back to the
chain head without trusting the operator or the registry.
"""

from __future__ import annotations

from typing import Any, Iterable, TypedDict

from algovoi_substrate.canonicalize import sha256_jcs


class AuditChainError(ValueError):
    """Raised when an audit chain row or sequence violates the discipline."""


class AuditChainRow(TypedDict):
    chain_position: int
    content_hash: str
    prev_hash: str | None
    payload: Any  # arbitrary JCS-canonicalisable payload


def _hash_payload(payload: Any) -> str:
    return sha256_jcs(payload)


def append_to_chain(
    *,
    payload: Any,
    prev_row: AuditChainRow | None,
) -> AuditChainRow:
    """Build the next row in an audit chain.

    If prev_row is None, the new row becomes the head:
        chain_position = 0, prev_hash = None
    Otherwise:
        chain_position = prev_row.chain_position + 1
        prev_hash = prev_row.content_hash

    The new row's content_hash is computed from the payload via JCS+SHA-256.
    """
    if prev_row is None:
        chain_position = 0
        prev_hash: str | None = None
    else:
        if not isinstance(prev_row.get("chain_position"), int) or isinstance(
            prev_row.get("chain_position"), bool
        ):
            raise AuditChainError(
                "prev_row.chain_position must be a non-negative integer"
            )
        if prev_row["chain_position"] < 0:
            raise AuditChainError(
                "prev_row.chain_position must be non-negative"
            )
        if not isinstance(prev_row.get("content_hash"), str) or not prev_row[
            "content_hash"
        ]:
            raise AuditChainError(
                "prev_row.content_hash must be a non-empty string"
            )
        chain_position = prev_row["chain_position"] + 1
        prev_hash = prev_row["content_hash"]

    return AuditChainRow(
        chain_position=chain_position,
        content_hash=_hash_payload(payload),
        prev_hash=prev_hash,
        payload=payload,
    )


def verify_audit_chain(rows: Iterable[AuditChainRow]) -> None:
    """Verify a chain of audit rows is well-formed and linkage-consistent.

    Checks:
    - chain_position starts at 0 and increments by exactly 1 per row.
    - prev_hash is None for the head row, otherwise equals the previous
      row's content_hash.
    - content_hash recomputes to SHA-256(JCS(payload)) of the carried
      payload.

    Raises AuditChainError on any violation; otherwise returns None.
    """
    expected_position = 0
    prev_content_hash: str | None = None

    for i, row in enumerate(rows):
        position = row.get("chain_position")
        if position != expected_position:
            raise AuditChainError(
                f"row {i}: chain_position {position} != expected "
                f"{expected_position}"
            )

        recomputed = _hash_payload(row.get("payload"))
        if row.get("content_hash") != recomputed:
            raise AuditChainError(
                f"row {i}: content_hash mismatch -- stored "
                f"{row.get('content_hash')!r} vs recomputed {recomputed!r}"
            )

        if expected_position == 0:
            if row.get("prev_hash") is not None:
                raise AuditChainError(
                    f"row 0 (head): prev_hash must be None, got "
                    f"{row.get('prev_hash')!r}"
                )
        else:
            if row.get("prev_hash") != prev_content_hash:
                raise AuditChainError(
                    f"row {i}: prev_hash {row.get('prev_hash')!r} "
                    f"!= previous content_hash {prev_content_hash!r}"
                )

        prev_content_hash = row["content_hash"]
        expected_position += 1

File: typescript/src/audit-chain.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Christopher Hopley / AlgoVoi (chopmob-cloud)
/**
 * Audit chain primitive: monotonic per-row hash chain.
 *
 * Each row carries:
 * - content_hash: SHA-256(JCS(payload))
 * - prev_hash: previous row's content_hash, or null for the head row
 * - chain_position: monotonically increasing integer starting at 0
 *
 * This is the structure AlgoVoi's production /compliance/attestation chain
 * exposes (Object Lock COMPLIANCE retention to 2033-05-04).
 */

import { sha256Jcs } from './canonicalize.js';

export class AuditChainError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'AuditChainError';
  }
}

export interface AuditChainRow {
  chain_position: number;
  content_hash: string;
  prev_hash: string | null;
  payload: unknown;
}

/**
 * Build the next row in an audit chain.
 *
 * If prevRow is null, the new row becomes the head (chain_position = 0,
 * prev_hash = null). Otherwise chain_position increments and prev_hash
 * links to prevRow.content_hash.
 */
export function appendToChain(
  payload: unknown,
  prevRow: AuditChainRow | null,
): AuditChainRow {
  let chainPosition: number;
  let prevHash: string | null;

  if (prevRow === null) {
    chainPosition = 0;
    prevHash = null;
  } else {
    if (
      typeof prevRow.chain_position !== 'number' ||
      !Number.isInteger(prevRow.chain_position) ||
      prevRow.chain_position < 0
    ) {
      throw new AuditChainError(
        'prevRow.chain_position must be a non-negative integer',
      );
    }
    if (
      typeof prevRow.content_hash !== 'string' ||
      prevRow.content_hash.length === 0
    ) {
      throw new AuditChainError(
        'prevRow.content_hash must be a non-empty string',
      );
    }
    chainPosition = prevRow.chain_position + 1;
    prevHash = prevRow.content_hash;
  }

  return {
    chain_position: chainPosition,
    content_hash: sha256Jcs(payload),
    prev_hash: prevHash,
    payload,
  };
}

/**
 * Verify a chain of audit rows is well-formed and linkage-consistent.
 *
 * Throws AuditChainError on any violation; otherwise returns void.
 */
export function verifyAuditChain(rows: readonly AuditChainRow[]): void {
  let expectedPosition = 0;
  let prevContentHash: string | null = null;

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i]!;

    if (row.chain_position !== expectedPosition) {
      throw new AuditChainError(
        `row ${i}: chain_position ${row.chain_position} != expected ${expectedPosition}`,
      );
    }

    const recomputed = sha256Jcs(row.payload);
    if (row.content_hash !== recomputed) {
      throw new AuditChainError(
        `row ${i}: content_hash mismatch -- stored ${row.content_hash} vs recomputed ${recomputed}`,
      );
    }

    if (expectedPosition === 0) {
      if (row.prev_hash !== null) {
        throw new AuditChainError(
          `row 0 (head): prev_hash must be null, got ${JSON.stringify(row.prev_hash)}`,
        );
      }
    } else {
      if (row.prev_hash !== prevContentHash) {
        throw new AuditChainError(
          `row ${i}: prev_hash ${JSON.stringify(row.prev_hash)} != previous content_hash ${JSON.stringify(prevContentHash)}`,
        );
      }
    }

    prevContentHash = row.content_hash;
    expectedPosition += 1;
  }
}
